rpc, wallet: Do not return "keypoololdest" for blank descriptor wallets

hebasto opened the pull request:
The "keypoololdest" field in the `getwalletinfo` RPC response should be used for legacy wallets only.

Th current implementation (04437ee721e66a7b76bef5ec2f88dd1efcd03b84) assumes that `CWallet::GetOldestKeyPoolTime()` always return `0` for descriptor wallets. This assumption is wrong for _blank_ descriptor wallets, when `m_spk_managers` is empty. As a result:
```
$ src/bitcoin-cli -signet -rpcwallet=211024-d-DPK getwalletinfo
{
  "walletname": "211024-d-DPK",
  "walletversion": 169900,
  "format": "sqlite",
  "balance": 0.00000000,
  "unconfirmed_balance": 0.00000000,
  "immature_balance": 0.00000000,
  "txcount": 0,
  "keypoololdest": 9223372036854775807,
  "keypoolsize": 0,
  "keypoolsize_hd_internal": 0,
  "paytxfee": 0.00000000,
  "private_keys_enabled": false,
  "avoid_reuse": false,
  "scanning": false,
  "descriptors": true
}
```

This PR fixes this issue with direct checking of the `WALLET_FLAG_DESCRIPTORS` flag.

lsilva01 reviewed (approved):
Code Review and Tested ACK 303ee60

nit: Maybe the comment in `DescriptorScriptPubKeyMan::GetOldestKeyPoolTime()` can be updated.

hebasto committed:
wallet, refactor: Make GetOldestKeyPoolTime return type std::optional

This change gets rid of the magic number 0 in the
DescriptorScriptPubKeyMan::GetOldestKeyPoolTime() function.

No behavior change.

hebasto committed:
wallet: Make GetOldestKeyPoolTime return nullopt for blank wallets

This change suppress the "keypoololdest" field in the getwalletinfo RPC
response for blank descriptor wallets.

hebasto commented:
Thanks to @lsilva01, @laanwj and @luke-jr for your reviews and suggestions!

Suggested an alternative implementation which is based on @laanwj's and @luke-jr's comments.

luke-jr commented on src/wallet/wallet.cpp:
```diff
@@ -2169,14 +2169,21 @@ bool CWallet::GetNewChangeDestination(const OutputType type, CTxDestination& des
     LOCK(cs_wallet);
-    int64_t oldestKey = std::numeric_limits<int64_t>::max();
+    std::optional<int64_t> oldest_key = std::nullopt;
     for (const auto& spk_man_pair : m_spk_managers) {
-        oldestKey = std::min(oldestKey, spk_man_pair.second->GetOldestKeyPoolTime());
+        if (!oldest_key.has_value()) {
+            oldest_key = spk_man_pair.second->GetOldestKeyPoolTime();
+            continue;
+        }
+        if (!spk_man_pair.second->GetOldestKeyPoolTime().has_value()) {
+            continue;
+        }
+        oldest_key = std::min(oldest_key.value(), spk_man_pair.second->GetOldestKeyPoolTime().value());
```
This is a behaviour change: Previously, it would return 0 if any of the m_spk_managers returned 0. With the new semantics, it should prefer std::nullopt over a value.

If this change is intentional, please document why.

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/23348#issuecomment-958736346).

hebasto commented:
Updated 324172ee4f3a974a694b44d4c812d7c9a8e752d9 -> da5f8663b858d9bf027143666483ea5e05373344 ([pr23348.02](https://github.com/hebasto/bitcoin/commits/pr23348.02) -> [pr23348.03](https://github.com/hebasto/bitcoin/commits/pr23348.03), [diff](https://github.com/hebasto/bitcoin/compare/pr23348.02..pr23348.03)):

- addressed @luke-jr's [comment](https://github.com/bitcoin/bitcoin/pull/23348#discussion_r741554082)

hebasto commented:
Updated da5f8663b858d9bf027143666483ea5e05373344 -> ee03c782ba61993d9e95fa499546cd14cee35445 ([pr23348.03](https://github.com/hebasto/bitcoin/commits/pr23348.03) -> [pr23348.04](https://github.com/hebasto/bitcoin/commits/pr23348.04), [diff](https://github.com/hebasto/bitcoin/compare/pr23348.03..pr23348.04)):

- split in two commits: a pure refactoring with no behavior change, and a bugfix.

luke-jr commented on src/wallet/wallet.cpp:
```diff
@@ -2169,14 +2169,18 @@ bool CWallet::GetNewChangeDestination(const OutputType type, CTxDestination& des
+    std::optional<int64_t> oldest_key{std::numeric_limits<int64_t>::max()};
     for (const auto& spk_man_pair : m_spk_managers) {
-        oldestKey = std::min(oldestKey, spk_man_pair.second->GetOldestKeyPoolTime());
+        oldest_key = std::min(oldest_key, spk_man_pair.second->GetOldestKeyPoolTime());
```
Couldn't you get a nullopt in here? (If not, why change this block at all?)

hebasto replied:
> Couldn't you get a nullopt in here?

Sorry, I did not get your question. What do you mean?

luke-jr replied:
`spk_man_pair.second->GetOldestKeyPoolTime()` could return `std::nullopt`. Is it well-defined what `std::min` will do here, and is it the desired behaviour?

hebasto replied:
[`std::min()`](https://en.cppreference.com/w/cpp/algorithm/min) uses `operator<`, which is [well-defined](https://en.cppreference.com/w/cpp/utility/optional/operator_cmp) for `std::optional`.

lsilva01 reviewed (approved):
re-ACK ee03c78

Tested `std::min()` with null value `std::optional <int64_t>` and it worked as expected.

stratospher reviewed (commented):
ACK ee03c78.
- [x] 3e4f069
    - The return type of all the `GetOldestKeyPoolTime()` is changed from `int64_t` to `optional<int64_t>`.
    - DescriptorScriptPubKeyMan’s `GetOldestKeyPoolTime()` returns `nullopt` instead of 0 and changes were made in `getwalletinfo()` to handle this.
- [x] ee03c78
    - In CWallet’s `GetOldestKeyPoolTime()`,  If m_spk_managers is not present, `nullopt` is returned(previously it returned `numeric_limits<int64_t>::max()`) and handles the case for blank descriptor wallets.

meshcollider reviewed (commented):
Code review ACK ee03c782ba61993d9e95fa499546cd14cee35445
